SourceX/qol.cpp

```diff
@@ -155,39 +190,103 @@ void DrawXPBar(CelOutputBuffer out)
 	if (!sgOptions.Gameplay.bExperienceBar)
 		return;
 
-	int barWidth = 306;
-	int barHeight = 5;
-	int yPos = gnScreenHeight - 9;                 // y position of xp bar
-	int xPos = (gnScreenWidth - barWidth) / 2 + 5; // x position of xp bar
-	int dividerHeight = 3;
-	int numDividers = 10;
-	int barColor = 198;
-	int emptyBarColor = 0;
-	int frameColor = 196;
-	bool space = true; // add 1 pixel separator on top/bottom of the bar
-
-	PrintGameStr(out, xPos - 22, yPos + 6, "XP", COL_WHITE);
-	int charLevel = plr[myplr]._pLevel;
-	if (charLevel == MAXCHARLEVEL - 1)
+	const PlayerStruct& player = plr[myplr];
+
+	const int backX = gnScreenWidth / 2 - 155;
+	const int backY = gnScreenHeight - 11;
+
+	const int xPos = backX + 3;
+	const int yPos = backY + 2;
+
+	DrawArt(out, backX, backY, &qolArt->xpbar);
+
+	const int charLevel = player._pLevel;
+
+	if (charLevel == MAXCHARLEVEL - 1) {
+		// Draw a nice golden bar for max level characters.
+		FastDrawHorizLine(out, xPos, yPos + 1, xpbar::barWidth, xpbar::goldGradient[xpbar::goldGrades * 3 / 4 - 1]);
+		FastDrawHorizLine(out, xPos, yPos + 2, xpbar::barWidth, xpbar::goldGradient[xpbar::goldGrades - 1]);
+		FastDrawHorizLine(out, xPos, yPos + 3, xpbar::barWidth, xpbar::goldGradient[xpbar::goldGrades / 2 - 1]);
+
 		return;
+	}
 
-	int prevXp = ExpLvlsTbl[charLevel - 1];
-	if (plr[myplr]._pExperience < prevXp)
+	const int prevXp = ExpLvlsTbl[charLevel - 1];
+	if (player._pExperience < prevXp)
 		return;
 
-	Uint64 prevXpDelta_1 = plr[myplr]._pExperience - prevXp;
-	int prevXpDelta = ExpLvlsTbl[charLevel] - prevXp;
-	int visibleBar = barWidth * prevXpDelta_1 / prevXpDelta;
+	Uint64 prevXpDelta_1 = player._pExperience - prevXp;
+	Uint64 prevXpDelta = ExpLvlsTbl[charLevel] - prevXp;

```

Comment:
Please avoid touching unrelated code. Uin64 is actually not the appropriate type to use here.

Reply:
The calculations I'm doing exceed SInt32 range. Rather than casting all the time, I'm using a sufficiently sized datatype for all involved variables.

Reply:
Subtracting 2 positive SInt32 doesn't need 64 :D

Reply:
Nope, but xpbar::barWidth * prevXpDelta_1 / prevXpDelta (right below) does. If all participants are integers, you get interesting results.

So either you cast (yuck) or you just make the types 64 bit wide by default, which seemed fine and less contentious. Guess I was wrong :D!